### ISSUE TITLE
loki.source.kubernetes_events fails to start with large number of events in a namespace ("event watcher exited with error")

### ISSUE THREAD

**Issue by towolf** [grafana/alloy#233]
### What's wrong?

We are using grafana-agent to ship Events off cluster and we are using a system that has a crazy pod churn.

The component worked fine initially, but as our churn increased it just stopped working after startup with:

"event watcher exited with error" 

It is timing out trying to start the informer. 

In the code you have this comment:

https://github.com/grafana/agent/blob/main/component/loki/source/kubernetes_events/kubernetes_events.go#L25-L26

I think for our use case it is not generous enough, becasue the initial sync is too heavy.

I would be cool if we could just start with new events or something and ignore the backlog, if that is at all possible.

### Steps to reproduce

Use the `loki.source.kubernetes_events` in a cluster with a very large amount of events.

Observe timeout error.

### System information

GKE 1.27 and RKE2 1.26

### Software version

docker.io/grafana/agent:v0.39.1

### Configuration

```text
logging {
      level = "debug"
    }

    //
    // Ship Kubernetes Events
    //
    loki.source.kubernetes_events "all" {
      forward_to = [loki.process.events.receiver]
      job_name = "integrations/kubernetes/eventhandler"
      // log_format = "logfmt"
      // namespaces = ["xyz"]
    }

    loki.process "events" {
      forward_to = [loki.write.central.receiver]

      stage.match {
        selector = "{namespace=~\"xzy\"}"

        stage.tenant {
          label = "namespace"
        }
      }

      stage.metrics {
        metric.counter {
          name        = "events_total"
          description = "total number of events"
          prefix      = "agent_"

          match_all         = true
          action            = "inc"
          max_idle_duration = "5m"
        }

        metric.counter {
          name        = "event_bytes_total"
          description = "total bytes of events"
          prefix      = "agent_"

          match_all         = true
          count_entry_bytes = true
          action            = "add"
          max_idle_duration = "5m"
        }
      }
    }

    loki.write "central" {
      endpoint {
        url = "https://xyz/loki/api/v1/push"
        tenant_id = "main"
        basic_auth {
          username = "loki"
          password = "xyz"
        }
      }
      external_labels = {
        cluster = "xyz",
      }
    }
```

### Logs

`ts=2024-02-25T20:35:47.698339436Z level=error msg="event watcher exited with error" component=loki.source.kubernetes_events.all err="failed to configure informers: timeout exceeded while configuring informers. Check the connection to the Kubernetes API is stable and that the Agent has appropriate RBAC permissions for &Event{ObjectMeta:{      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] [] []},InvolvedObject:ObjectReference{Kind:,Namespace:,Name:,UID:,APIVersion:,ResourceVersion:,FieldPath:,},Reason:,Message:,Source:EventSource{Component:,Host:,},FirstTimestamp:0001-01-01 00:00:00 +0000 UTC,LastTimestamp:0001-01-01 00:00:00 +0000 UTC,Count:0,Type:,EventTime:0001-01-01 00:00:00 +0000 UTC,Series:nil,Action:,Related:nil,ReportingController:,ReportingInstance:,}"`
`ts=2024-02-25T20:35:47.698440294Z level=info msg="stopping watcher for events" `

**Comment by aglees** [comment]
I think we're experiencing the same issue, and this could well be a limiter in some of our larger clusters (>50 nodes) with a good amount of churn. This wasn't an issue AFAIK with Grafana Agent.

**Comment by aglees** [comment]
@towolf are you able to create a PR for that change to the timeout? Do you think this could be set as a command-line parameter?

**Comment by towolf** [comment]
I could use input from the maintainers on whether they would accept this simple fix and whether it needs to be confiurable.

**Comment by nrwiersma** [comment]
Took a quick look at the code for this, and if configurability were needed, I would likely add it to the [Arguments](https://github.com/grafana/alloy/blob/main/internal/component/loki/source/kubernetes_events/kubernetes_events.go#L43) and likely fix `podlogs` the same way while I was at it.

If this works as a solution, could open a PR next week.

**Comment by rfratto** [comment]
Preferably we avoid new configuration settings whenever possible, since every new configuration setting becomes part of our backwards compatibility guarantees. If we were going to add a new configuration syntax, @nrwiersma has the right idea of adding it to the Arguments struct. 

It feels like a good path forward for now would be to make that hard-coded timeout significantly more generous. I wrote the original code and picked 5s on a whim, and didn't consider what might've been needed for larger clusters. 

I'm not quite sure how much more generous the timeout needs to be, but I think we can adjust it one or two more times before considering making it configurable. Since we're just talking about a timeout, I don't anticipate changing the timeout for all users to have any adverse effects. 

A PR from someone who is experiencing this issue that can help test an increased timeout would be very welcome.

**Comment by nrwiersma** [comment]
Good to know. When we made a custom build, we just set it to 10m, which is way to long, but unblocked us. Perhaps we can do some tests to find a more reasonable timeout extension.

**Comment by towolf** [comment]
The initial sync of the informer was definitely on the order of minutes. That's because in some namespaces we had hundreds of thousands of events. 

So to skip the backlog entirely would actually be preferable, ...

**Comment by aglees** [comment]
My 2c would be to parameterise the skipping of the backlog too.

**Comment by rfratto** [comment]
I merged #1088 for a short-term fix, but I recognize that even 10m may not be enough for sufficiently large clusters. If others continue to run into this after we make a release which includes #1088, please create a new issue (we don't monitor activity on older issues) and we'll discuss potential medium- or long-term fixes, including skipping the backlog.

**Comment by towolf** [comment]
Great. Thank you!

**Comment by aglees** [comment]
In terms of release expectations, do you see this as being a candidate for `v1.2.0`?

**Comment by rfratto** [comment]
> @rfratto: in terms of release expectations, do you see this as being a candidate for v1.2.0?

I think the window for v1.2.0 closed a while ago. This is either a candidate for v1.2.1 after we have a slightly bigger pile of bug fixes, or v1.3.0 at the worst case.
